src/txn/txn.c

```diff
@@ -253,18 +279,47 @@ __txn_oldest_scan(WT_SESSION_IMPL *session,
 
 	/* The oldest ID cannot change while we are holding the scan lock. */
 	prev_oldest_id = txn_global->oldest_id;
-	last_running = oldest_id = txn_global->current;
+	current_id = last_running = oldest_id = txn_global->current;
 	if ((metadata_pinned = txn_global->checkpoint_state.id) == WT_TXN_NONE)
 		metadata_pinned = oldest_id;
 
 	/* Walk the array of concurrent transactions. */
 	WT_ORDERED_READ(session_cnt, conn->session_cnt);
 	for (i = 0, s = txn_global->states; i < session_cnt; i++, s++) {
-		/* Update the last running transaction ID. */
-		if ((id = s->id) != WT_TXN_NONE &&
+		/* Update the last running transaction ID.
+		 *
+		 * Ignore:
+		 *  - The ID if it is higher than the current ID we saw. This
+		 *    can happen if the transaction is quickly done its job.
+		 *    In this case, we ignore this transaction since it would
+		 *    not be visible anyway in current snapshot.
+		 */
+		while ((id = s->id) != WT_TXN_NONE &&
 		    WT_TXNID_LE(prev_oldest_id, id) &&
-		    WT_TXNID_LT(id, last_running))
-			last_running = id;
+		    WT_TXNID_LT(id, last_running) &&
+		    WT_TXNID_LT(id, current_id)) {
```

Comment:
I think this line (comparing `id` with `current_id`) doesn't necessary. `last_running` value starts from the `current_id`(line 282), and `last_running` value is monotonically decreased in this loop. So if the `id` value is less than `last_running` value (line 299), it must be less than `current_id` value.